### DISCUSSION TITLE
Question about the implementation of Disposer.

### DISCUSSION THREAD

**Discussion by balusch** [capnproto/capnproto#1951] [Q&A] [answered]
Hi, I'm learning the use and implementation of the kj library, and find something hard to understand in the implementation of `kj::Disposer`. Specifically, why `dynamic_cast` is required while converting a `T *` to `void *` in the code piece below? 

https://github.com/capnproto/capnproto/blob/64802802df1d7780625eeb07b71d249fe49fb68d/c%2B%2B/src/kj/memory.h#L662-L670

Isn't it safe to convert all pointers to `void *` explicitly or implicitly? Can it be replaced by `static_cast`?

**Comment by kentonv** [comment, marked as answer]
`dynamic_cast<void*>(ptr)` is equivalent to down-casting the pointer to its most-derived type, then casting it to void*. This does not necessarily produce the exact same pointer value, particularly in the case of multiple inheritance.

```
class Bar { int i; }
class Baz { int j; }
class Foo: public Bar, public Baz {};

Foo foo;
Bar* bar = &foo;
Baz* baz = &foo;

// The pointers `bar` and `baz` are not exactly the same numerically!
assert(reinterpret_cast<void*>(bar) != reinterpret_cast<void*>(baz));

// But dynamic_cast will return the same value for both.
assert(dynamic_cast<void*>(bar) == dynamic_cast<void*>(baz));
assert(dynamic_cast<void*>(bar) == reinterpret_cast<void*>(&foo));
assert(dynamic_cast<void*>(baz) == reinterpret_cast<void*>(&foo));
```

The reason this is important for disposers is because an `Own<Foo>` could have been converted to `Own<Baz>`, but the disposer needs to receive the original pointer, not the pointer to the inherited object.

**Reply by balusch** [reply]
Though I haven't used multiple inheritance and `dynamic_cast` before, your explanation still looks clear and understandable. Thanks!

After some investigations of `dynamic_cast`, I learned that polymorphism is required to perform a downcast, which was omitted in your example. And then I began to think what if polymorphism does not exist and I still want to do such casting? Can the other specialization utilizing `static_cast` still work correctly in this case? And I found this casting had already been blocked. haha... Well done! 👍 

```c++
{ // polymorphism
  class Bar { int i; virtual void hello() {} };
  class Baz { int j; virtual void world() {} };
  class Foo : public Bar, public Baz {};
  kj::Own<Foo> foo = kj::heap<Foo>();
  kj::Own<Bar> bar = kj::mv(foo);
}

{ // non-polymorphism
  class Bar { int i; };
  class Baz { int j; };
  class Foo : public Bar, public Baz {};
  kj::Own<Foo> foo = kj::heap<Foo>();
  kj::Own<Bar> bar = kj::mv(foo); // WON'T compile!!!
}
```
